sdk/schemaregistry/azure-schemaregistry-client/src/main/java/com/azure/schemaregistry/client/CachedSchemaRegistryClient.java

```diff
@@ -0,0 +1,277 @@
+    static final int MAX_SCHEMA_MAP_SIZE_MINIMUM = 10;
+
+    private final AzureSchemaRegistryRestService restService;
+    private final int maxSchemaMapSize;
+    private final HashMap<String, Function<String, Object>> typeParserDictionary;
+    private final HashMap<String, SchemaRegistryObject> guidCache;
+    private final HashMap<String, SchemaRegistryObject> schemaStringCache;
+
+    CachedSchemaRegistryClient(
+        String registryUrl,
+        HttpPipeline pipeline,
+        int maxSchemaMapSize,
+        HashMap<String, Function<String, Object>> typeParserDictionary) {
+        if (registryUrl == null || registryUrl.isEmpty()) {
+            throw logger.logExceptionAsError(
+                new IllegalArgumentException("Schema Registry URL cannot be null or empty."));
+        }
+
+        this.restService = new AzureSchemaRegistryRestServiceClientBuilder()
+            .host(registryUrl)
+            .pipeline(pipeline)
+            .buildClient();
+
+        this.maxSchemaMapSize = maxSchemaMapSize;
+        this.typeParserDictionary = typeParserDictionary;
+        this.guidCache = new HashMap<>();
+        this.schemaStringCache = new HashMap<>();
+    }
+
+    // testing
+    CachedSchemaRegistryClient(
+        AzureSchemaRegistryRestService restService,
+        HashMap<String, SchemaRegistryObject> guidCache,
+        HashMap<String, SchemaRegistryObject> schemaStringCache,
+        HashMap<String, Function<String, Object>> typeParserDictionary) {
+        this.restService = restService; // mockable
+        this.guidCache = guidCache;
+        this.schemaStringCache = schemaStringCache;
+        this.typeParserDictionary = typeParserDictionary;
+        this.maxSchemaMapSize = MAX_SCHEMA_MAP_SIZE_DEFAULT;
+    }
+
+    /**
+     * @return Azure Schema Registry service string encoding
+     */
+    @Override
+    public Charset getEncoding() {
+        return CachedSchemaRegistryClient.SCHEMA_REGISTRY_SERVICE_ENCODING;
+    }
+
+    /**
+     * @param schemaType tag used by schema registry store to identify schema serialization type, e.g. "avro"
+     * @param parseMethod function to parse string into usable schema object
+     * @throws IllegalArgumentException on bad schema type or if parser for schema type has already been registered
+     */
+    public synchronized void loadSchemaParser(String schemaType, Function<String, Object> parseMethod) {
+        if (schemaType == null || schemaType.isEmpty()) {
+            throw logger.logExceptionAsError(
+                new IllegalArgumentException("Serialization type cannot be null or empty."));
+        }
+        if (this.typeParserDictionary.containsKey(schemaType.toLowerCase(Locale.ENGLISH))) {
+            throw logger.logExceptionAsError(
+                new IllegalArgumentException("Multiple parse methods for single serialization type may not be added."));
+        }
+        this.typeParserDictionary.put(schemaType.toLowerCase(Locale.ENGLISH), parseMethod);
```

Comment:
Yeah I think I am going to make all three ConcurrentHashMaps so I can remove `synchronized`

In this case I prefer to throw if multiple parsers are added for the same serialization type, just because upserting could potentially result in a functioning app with unexpected behaviors.